### PR TITLE
fix for detectors with sigma aggregation rules

### PR DESCRIPTION
**Pull request by sbcd90**
### Description
fix for detectors with sigma aggregation rules

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [X] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [X] Commits are signed per the DCO using `--signoff`.
- [X] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/security-analytics/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java [hunk 1]
```diff
@@ -2153,6 +2157,109 @@ public void testCreateDetectorWithCloudtrailAggrRuleWithEcsFields() throws IOExc
         assertEquals(1, getFindingsBody.get("total_findings"));
     }
 
+    @SuppressWarnings("unchecked")
+    public void testCreateDetectorWithCloudtrailAggrRuleWithRolloverIndexAliases() throws IOException, InterruptedException {
+        updateClusterSetting("plugins.security_analytics.enable_detectors_with_dedicated_query_indices", "false");
```
Comment (eirsep) [inline src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java:2162]: why do we turn off this setting?

Comment (eirsep) [reply src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java:2162]: plz change to true as that's the new default behaviour

File: src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java [hunk 2]
```diff
@@ -2153,6 +2157,109 @@ public void testCreateDetectorWithCloudtrailAggrRuleWithEcsFields() throws IOExc
...
+        params.put("detector_id", createdId);
+        Response getFindingsResponse = makeRequest(client(), "GET", SecurityAnalyticsPlugin.FINDINGS_BASE_URI + "/_search", params, null);
+        Map<String, Object> getFindingsBody = entityAsMap(getFindingsResponse);
+
+        // Assert findings
+        assertNotNull(getFindingsBody);
+        assertEquals(1, getFindingsBody.get("total_findings"));
+
+        doRollover("ocsf_ct");
+        Thread.sleep(90000);
```
Comment (eirsep) [inline src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java:2247]: why should we do sleep?
after rollover is the template not applied to new index directly?

File: src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java [hunk 3]
```diff
@@ -2153,6 +2157,109 @@ public void testCreateDetectorWithCloudtrailAggrRuleWithEcsFields() throws IOExc
...
+
+        indexDoc("ocsf_ct", "4", randomCloudtrailOcsfDoc());
```
Comment (eirsep) [inline src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java:2249]: before this should we not verify mappings of new write index?

**Review by eirsep — COMMENTED** [review]
when is template applied? not at creation time? why is there sleep 90 s in test?

File: src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java (hunk 2 above)
Comment (sbcd90) [reply src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java:2247]: The Sigma Aggregation Rule we're using has timeframe set to `1m`(https://github.com/opensearch-project/security-analytics/blob/622d26eb5ddf3cb03f22f1cf0e0cb65a8a5c04cf/src/test/java/org/opensearch/securityanalytics/TestHelpers.java#L1162). So, the `sleep` ensures that the rule only considers docs from the rolled over new write index. If this fix is not present, the `alerting workflow run` doesn't generate the new finding. https://github.com/opensearch-project/security-analytics/blob/622d26eb5ddf3cb03f22f1cf0e0cb65a8a5c04cf/src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java#L2251

File: src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java (hunk 3 above)
Comment (sbcd90) [reply src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java:2249]: mappings of the first write index are verified with the first workflow run. https://github.com/opensearch-project/security-analytics/blob/622d26eb5ddf3cb03f22f1cf0e0cb65a8a5c04cf/src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java#L2235 .If alias mappings are not applied, findings will not be generated.

**Comment by sbcd90** [comment]
> when is template applied? not at creation time? why is there sleep 90 s in test?

Yes. Template is applied at creation time with the `Create Mapping` api call.

File: src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java (hunk 1 above)
Comment (sbcd90) [reply src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java]: yes. sorry. leftover from some experiments.

**Review by eirsep — APPROVED** [review]

File: src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java (hunk 3 above)
Comment (eirsep) [reply src/test/java/org/opensearch/securityanalytics/resthandler/DetectorMonitorRestApiIT.java:2248]: But i thought the bug was that AFTER rollover the template is not getting applied?

File: src/main/java/org/opensearch/securityanalytics/mapper/IndexTemplateManager.java [hunk 4]
```diff
@@ -115,13 +115,22 @@ public void upsertIndexTemplateWithAliasMappings(
 
         upsertComponentTemplateStepListener.whenComplete( acknowledgedResponse -> {
 
-            // Find template which matches input index best
+            // Find template which matches input index best. starts by directly matching with input index and
```
Comment (riysaxen-amzn) [inline src/main/java/org/opensearch/securityanalytics/mapper/IndexTemplateManager.java:118]: let's use doc comments for these lines instead -> https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html

Comment (sbcd90) [reply src/main/java/org/opensearch/securityanalytics/mapper/IndexTemplateManager.java:118]: will address this in a follow-up pr.

**Review by riysaxen-amzn — APPROVED** [review]

**Pull request merged**